plugins/CoreAdminHome/Commands/InvalidateReportData.php

```diff
@@ -0,0 +1,272 @@
+            }
+
+            // validate dates
+            $startDate = $this->validateDate($parts[0]);
+            $endDate = $this->validateDate($parts[1]);
+
+            $result[] = array($startDate, $endDate);
+        }
+        return $result;
+    }
+
+    /**
+     * @param string[] $periodTypes
+     * @param Date[][] $dateRanges
+     * @param bool $cascade
+     * @return Date[][]
+     */
+    public function getPeriodsToInvalidateFor($periodTypes, $dateRanges, $cascade)
+    {
+        $result = array();
+
+        foreach ($dateRanges as $range) {
+            /**
+             *  @var Date $startDate
+             *  @var Date $endDate
+             */
+            list($startDate, $endDate) = $range;
+
+            foreach ($periodTypes as $type) {
+                /** @var Range $periodObj */
+                $periodObj = PeriodFactory::build($type, $startDate . ',' . $endDate);
+
+                $subperiods = $periodObj->getSubperiods();
+                foreach ($subperiods as $subperiod) {
+                    $result[$type][] = $subperiod->getDateStart()->toString();
+                }
+
+                if ($cascade) {
+                    $realStartDate = reset($subperiods)->getDateStart()->toString();
+                    $realEndDate = end($subperiods)->getDateEnd()->toString();
+
+                    $this->addPeriodsToCascadeOn($result, $type, $realStartDate, $realEndDate);
+                }
+            }
+        }
+
+        foreach ($result as $periodType => $dates) {
+            $result[$periodType] = array_unique($dates);
+        }
+
+        return $result;
+    }
+
+    /**
+     * TODO
+     *
+     * @param Date[][] $result
+     * @param string $periodType
+     * @param string $startDate
+     * @param string $endDate
+     */
+    private function addPeriodsToCascadeOn(&$result, $periodType, $startDate, $endDate)
+    {
+        $childPeriodType = $this->getChildPeriod($periodType);
+        if (empty($childPeriodType)) {
+            return;
+        }
+
+        $childPeriods = PeriodFactory::build($childPeriodType, $startDate . ',' . $endDate);
+
+        $subperiods = $childPeriods->getSubperiods();
+        foreach ($subperiods as $childPeriod) {
+            $result[$childPeriod->getLabel()][] = $childPeriod->getDateStart()->toString();
+        }
+
+        /** @var Date $realStartDate */
+        $realStartDate = reset($subperiods)->getDateStart();
+        if ($realStartDate->isEarlier(Date::factory($startDate))) {
+            $startDate = $realStartDate->toString();
+        }
+
+        /** @var Date $realEndDate */
+        $realEndDate = end($subperiods)->getDateEnd();
+        if ($realEndDate->isLater(Date::factory($endDate))) {
+            $endDate = $realEndDate->toString();
+        }
+
+        $this->addPeriodsToCascadeOn($result, $childPeriodType, $startDate, $endDate);
+    }
+
+    private function getChildPeriod($periodType)
```

Comment:
Maybe this could be moved into a period class? A `switch` is most times a sign that one needs subclasses for each entity. So each period class could have a method `getChildPeriod()`. So you get the Period class for your period via the factory and then call `$period->getChildPeriod()`